milavision/milavision/datasets.py

```diff
@@ -0,0 +1,122 @@
+"""Drop-in replacement for the `torchvision.datasets` package, optimized for Mila/CC clusters.
+
+When running locally, there is no difference between using this package and torchvision.
+
+When running on the Mila cluster, the only difference is that the `root` and `download` arguments
+that are normally passed to the dataset class (e.g. `MNIST(root, download=True)`) could be ignored.
+ 
+>>> # from torchvision.datasets import MNIST
+>>> from milavision.datasets import MNIST
+>>> dataset = MNIST("~/my/data/directory", download=True)
+>>> # dataset.root might not actually be "~/my/data/directory"!
+"""
+import functools
+import typing
+
+import torchvision.datasets as _tvd
+
+# Import EVERYTHING from torchvision, and then overwrite whatever we support.
+from torchvision.datasets import *  # type: ignore
+from torchvision.datasets import __all__
+
+from milavision._utils import ClusterType
+
+for attribute, value in vars(_tvd).items():
+    if attribute.startswith("__") and attribute.endswith("__"):
+        continue
+    if attribute not in __all__:
+        """
+        For all the attributes like `torchvision.datasets.vision`, `torchvision.datasets.utils`,
+        or the modules like `torchvision.datasets.caltech`, etc, we copy over the value to the
+        globals here.
+        This makes it possible to do:
+        
+        ```python
+        from milavision.datasets import caltech
+        ```
+        """
+        globals()[attribute] = value
+
+if typing.TYPE_CHECKING:
+    # Import these here, so that when writing code, you get hints when doing something like:
+    #
+    # from milavision.datasets import caltech
+    #
+    # without the type-checker complaining.
```

Comment:
I mean, if you're going to paste the whole list of imports here regardless, why not just do it unconditionally and get rid of the logic above it? If we are worried about missing things in newer versions of torchvision, this can be tested for and checked by the CI.

Now this is a bit of a philosophical point, but I think it would be better to explicitly not allow importing datasets from milavision that we do not support. It's expectation management: if I can import something from milavision, I expect it to do the optimal thing on the Mila cluster. If it turns out we don't support it, I think I would prefer to get an exception. Then I know that okay I have to use the default version and I have to be careful.

Do tell if you see use cases where it is important to have full coverage within the same import.